### PR TITLE
REPL tab completion for non-struct types

### LINKED ISSUE #41652: REPL lacks tab completion for non-struct types
Defining `propertynames` for non-struct types is not enough to enable tab completion in the REPL, as is being done with `Cptr` (https://github.com/analytech-solutions/CBinding.jl/issues/77) and eventually with `Ptr` (https://github.com/JuliaInterop/Clang.jl/issues/316).  It appears to be prevented because of the following `isstructtype` check: https://github.com/JuliaLang/julia/blob/c95448fb988441ecd7fb0604606ac08c5014abd3/stdlib/REPL/src/REPLCompletions.jl#L147

And indeed, defining something like `Base.isstructtype(::Type{<:Cptr}) = true` gets tab completion working, but is fairly a cringe-worthy workaround.  Is this a bug in (or feature request for) the REPL?  Or is there another approach besides the `isstructtype` workaround?

### PR DESCRIPTION
**Pull request by ArunS-tack**
closes #41652.

### CODE REVIEW

**Comment by KristofferC** [comment]
Would be good with a test as well

**Comment by ArunSanganal** [comment]
Not quite sure what or how to add a test case for REPL tab completion

**Comment by KristofferC** [comment]
You can see an example test here:

https://github.com/JuliaLang/julia/blob/1c951f7fe8e5e923cc5f5b18c91e461b61641b14/stdlib/REPL/test/replcompletions.jl#L1078-L1081

**Comment by ArunSanganal** [comment]
Ok, I got how to add a test case but I am still confused about a good test, I was thinking of adding the example mentioned in the issue but it uses another package.

**Comment by KristofferC** [comment]
Perhaps you can extract a minimal working example from there which didn't work before this PR but now works.

**Comment by krrutkow** [comment]
Here is a MWE to demonstrate the issue without a package:
```julia
julia> primitive type NonStruct 8 end

julia> Base.propertynames(::NonStruct) = (:a, :b, :c)

julia> x = reinterpret(NonStruct, 0x00)
NonStruct(0x00)

julia> x.  # tab complete
!                        InteractiveUtils          conj                      ismissing                 replace!
!=                       InterruptException        conj!                     ismount                   repr
!==                      InvalidStateException     contains                  ismutable                 reset
⋮

julia> Base.isstructtype(::Type{NonStruct}) = true

julia> x.  # tab complete
a b  c
```

**Comment by ArunSanganal** [comment]
Weird, the test is passing for me locally but not in here.

**Comment by ArunSanganal** [comment]
Test is passing locally for me but not in here. Any idea why is that?

**Comment by KristofferC** [comment]
Hm, perhaps try move the `@eval` block outisde the `@testset`, (so the eval is not needed)

**Comment by ArunSanganal** [comment]
> Hm, perhaps try move the `@eval` block outisde the `@testset`, (so the eval is not needed)

still not passing :/

File: stdlib/REPL/test/replcompletions.jl [hunk 1]
```diff
@@ -1145,6 +1145,14 @@ let s = "test_dict[\"ab"
     @test c == Any["\"abc\"", "\"abcd\""]
 end
 
+primitive type NonStruct 8 end
+Base.propertynames(::NonStruct) = (:a, :b, :c)
+x = reinterpret(NonStruct, 0x00)
```
Comment (vtjnash) [inline stdlib/REPL/test/replcompletions.jl:1150]: Move this into `module CompletionFoo` at the top of the file. The REPL runs in module Main, while here the test runs in an anonymous module.

(and make the test for `s = "CompletionFoo.x."`)

**Review by vtjnash — APPROVED** [review]

**Pull request merged**